api/src/test/java/org/pmiops/workbench/access/UserServiceAccessTest.java

```diff
@@ -170,131 +178,380 @@ public void test_updateUserWithRetries_register_includes_others() {
+        user -> {
+          user.setDataUseAgreementBypassTime(null);
+          user.setDataUseAgreementCompletionTime(Timestamp.from(START_INSTANT));
+          return userDao.save(user);
+        });
+  }
+
+  @Test
+  public void test_updateUserWithRetries_dua_unbypassed_no_aar_wrong_version_noncompliant() {
+    providedWorkbenchConfig.access.enableDataUseAgreement = true;
+    providedWorkbenchConfig.access.enableAccessRenewal = false;
+    assertUnregisters(
+        user -> {
+          user.setDataUseAgreementBypassTime(null);
+          user.setDataUseAgreementCompletionTime(Timestamp.from(START_INSTANT));
+          user.setDataUseAgreementSignedVersion(userService.getCurrentDuccVersion() - 1);
+          return userDao.save(user);
+        });
+  }
+
+  @Test
+  public void test_updateUserWithRetries_dua_unbypassed_aar_noncompliant() {
+    providedWorkbenchConfig.access.enableDataUseAgreement = true;
+    providedWorkbenchConfig.access.enableAccessRenewal = true;
+    providedWorkbenchConfig.accessRenewal.expiryDays = 365L;
+    assertUnregisters(
+        user -> {
+          user.setDataUseAgreementBypassTime(null);
+          return userDao.save(user);
+        });
+  }
+
+  @Test
+  public void test_updateUserWithRetries_dua_unbypassed_aar_missing_version_noncompliant() {
+    providedWorkbenchConfig.access.enableDataUseAgreement = true;
+    providedWorkbenchConfig.access.enableAccessRenewal = true;
+    providedWorkbenchConfig.accessRenewal.expiryDays = 365L;
+    assertUnregisters(
+        user -> {
+          user.setDataUseAgreementBypassTime(null);
+          user.setDataUseAgreementCompletionTime(Timestamp.from(START_INSTANT));
+          return userDao.save(user);
+        });
+  }
+
+  @Test
+  public void test_updateUserWithRetries_dua_unbypassed_aar_wrong_version_noncompliant() {
+    providedWorkbenchConfig.access.enableDataUseAgreement = true;
+    providedWorkbenchConfig.access.enableAccessRenewal = true;
+    providedWorkbenchConfig.accessRenewal.expiryDays = 365L;
+    assertUnregisters(
+        user -> {
+          user.setDataUseAgreementBypassTime(null);
+          user.setDataUseAgreementCompletionTime(Timestamp.from(START_INSTANT));
+          user.setDataUseAgreementSignedVersion(userService.getCurrentDuccVersion() - 1);
+          return userDao.save(user);
+        });
+  }
+
+  @Test
+  public void test_updateUserWithRetries_dua_unbypassed_aar_expired_noncompliant() {
+    providedWorkbenchConfig.access.enableDataUseAgreement = true;
+    providedWorkbenchConfig.access.enableAccessRenewal = true;
+
+    final long expirationWindow = 365L;
+    providedWorkbenchConfig.accessRenewal.expiryDays = expirationWindow;
+    final Timestamp expired =
+        Timestamp.from(START_INSTANT.minus(expirationWindow + 1, ChronoUnit.DAYS));
+
+    assertUnregisters(
+        user -> {
+          user.setDataUseAgreementBypassTime(null);
+          user.setDataUseAgreementCompletionTime(expired);
```

Comment:
For these expiration tests, it's probably more accurate to set the expired timestamp up front, and then clear the bypass. As currently written, the test will simulate setting a date over a year in the past, which would never happen in practice.

Mainly this is just from a black box / separation of concerns standpoint. I realize what I described is effectively equivalent to what you have written here, knowing the implementation.

I guess it will be difficult to apply this pattern if there's no bypass, e.g. for the publish/profile timestamps, unless you advance the clock within the callback (that may be too complicated to be worth it).